addons/base_import/models/base_import.py

```diff
@@ -476,107 +554,208 @@ def _try_match_date_time(self, preview_values, options):
+            :param (list) header_types: Allowed types for each column in the parsed file.
+                Coming from "_get_most_likely_header_types" method
+                E.g.: ['int', 'float', 'char', 'many2one', ...]
+            :returns: if the header couldn't be matched : an empty dict
+                      else: a dict with the field path and the distance between header and the matched field.
+            :rtype: dict(field_path + Levenshtein distance)
+                    In case of simple matching : {'field_path': [field_name], distance: Levenshtein_distance}
+                                           e.g.: {'field_path': ['lead_id'], distance: 0.23254}
+                    In case of hierarchy matching : {'field_path': [parent_field_name, child_field_name, subchild_field_name]}
+                                              e.g.: {'field_path': ['lead_id', 'description']}
         """
-        string_match = None
-        IrTranslation = self.env['ir.translation']
-        for field in fields:
-            # FIXME: should match all translations & original
-            # TODO: use string distance (levenshtein? hamming?)
-            if header.lower() == field['name'].lower():
-                return [field]
-            if header.lower() == field['string'].lower():
-                # matching string are not reliable way because
-                # strings have no unique constraint
-                string_match = field
-            translated_header = IrTranslation._get_source('ir.model.fields,field_description', 'model', self.env.lang, header).lower()
-            if translated_header == field['string'].lower():
-                string_match = field
-        if string_match:
-            # this behavior is only applied if there is no matching field['name']
-            return [string_match]
+        if not fields_tree:
+            return {}
 
         if '/' not in header:
-            return []
+            # First, try exact match
+            IrTranslation = self.env['ir.translation']
+            for field in fields_tree:
+                # FIXME: should match all translations & original
+                string_match = None
+                if header.lower() == field['name'].lower():
+                    return {
+                        'field_path': [field['name']],
+                        'distance': 0
+                    }
+                field_string = field.get('string', '').lower()
+                if header.lower() == field_string:
+                    # matching string are not reliable way because
+                    # strings have no unique constraint
+                    string_match = field
+                translated_header = IrTranslation._get_source('ir.model.fields,field_description', 'model', self.env.lang, header).lower()
+                if translated_header == field_string:
+                    string_match = field
```

Comment:
It's a bit hard to see which case is what and which has precedence. 

Why not:

```py
field_match = False
for field in fields_tree:
    if header.lower() == field['name'].lower():
        field_match = field  # found an exact match based on the field technical name

    field_string = field.get('string', '')
    translated_header = IrTranslation._get_source('ir.model.fields,field_description', 'model', self.env.lang,  header).lower()
    if translated_header.lower() == field_string.lower():  # first, try to match using user translation
        field_match = field  # found an exact match based on translated field label
        break
    elif header.lower() == field_string.lower():  # then, try to match raw English term
        field_match = field  # found an exact match based on raw English field label
        break

if field_match:  # found an exact match, no need to go further
    return {
        'field_path': [field_match['name']],
        'distance: 0' # exact match has a 0 distance
    }
```

Reply:
reminiscence of the past, once again :D